### ISSUE TITLE
Autocomplete for PostgreSQL search backend

### ISSUE THREAD

**Issue by kaedroho** [wagtail/wagtail#3709]
Autocomplete isn't currently supported by the PostgreSQL backend (#3515). We've had to switch a couple of clients back to Elasticsearch because the search experience isn't as good without it.

I've made a start of implementing this in https://github.com/kaedroho/wagtail/tree/feature/postgres-search-autocomplete. It's still early days though, just opening this issue to track progress and make other developers aware of what I'm working on.

**Comment by BertrandBordage** [comment]
Sorry to be so direct, but I don’t think this is the right way to implement it.
You’re basically creating explicitly in a new column what’s already in the GIN index. Except that unfortunately, it’ll be less efficient than GIN.

As you can see [in the PostgreSQL docs](https://www.postgresql.org/docs/9.6/static/textsearch-controls.html), it is possible to do “prefix matching” using a syntax such as `to_tsquery('supercalif:*')`.

I already implemented it [in django-postgrefts](https://github.com/BertrandBordage/django-postgrefts/blob/master/postgrefts/models.py#L194-L201), it works very well, but I don’t think we can implement it properly in wagtail without doing some preliminary API changes. That’s because we don’t really need to create a new column/index just for prefix matching a field.

That’s one of the reasons why I told you a few months ago that we should discuss about changing the search API ;) One of the main others reasons is multi-language support of search, as the multi-index is not at all the most efficient solution, at least for PostgreSQL.

**Comment by kaedroho** [comment]
@BertrandBordage Wagtail at the moment lets you specify which ``SearchFields`` have ``partial_match`` enabled. So to make it behave properly, we have to index those fields separately (but, of course, we can discuss changing that behaviour).

I was also thinking that precalculating the edgengrams would be slightly faster as each term in the search query would map to a term in the database whereas prefix queries may have to pull out many terms and combine the results on the fly. AFAIK, there isn't a reason why two GIN indexes can't be used simultaneously in the same query (but that isn't documented so I don't know this for sure).

**Comment by BertrandBordage** [comment]
Well, matching a prefix and using a binary tree index (or a GIN index, GIN being an improved b-tree index) are exactly the same thing: if you search for “something”, a first step is made into the tree, at the “s” root, then we continue to the “o” node, etc until the “g” leaf. On each step the number of results is narrowed down. So when you do a prefix matching with a database index, you actually do less work (results fetching put aside) than to do an full match.

By creating a field of what Elasticsearch calls edge Ngrams, you actually create exactly what’s inside an index, except that it’s not sorted since it’s not an index, so it will be way slower to search for your term than if you searched directly in the original database columns, which is of course way slower than searching in a unified index.

Just creating a column compiling all `partial_match`ed `SearchField`s without this tokenization but with a GIN index would be way faster for roughly the same table size.

But again, this is far from ideal compared to what we aim to achieve: having the same search functionalities with both Elasticsearch & PostgreSQL. If we want to be able to search for a single column or method, with or without partial match, the current PostgreSQL system I made is not enough.

Ideally, here is what I think the PostgreSQL storage should look like: all `SearchField`s (so database columns + methods) should be stored independently, so we could search on a given field, with or without autocomplete. I have multiple ideas to make this a reality, but it’s quite tricky to write. The most obvious solution is to use an column containing an array of `ts_vector`s, combined with a strong Python introspection system to know which item is which column, and also to be able to search without specifying a column. I have other ideas like using materialized views containing ts_vectors for each page table, all combined by a view being the overall index. It could work very well, but given the Django support for these technologies, I want to avoid it.

**Comment by kaedroho** [comment]
> Well, matching a prefix and using a binary tree index (or a GIN index, GIN being an improved b-tree index) are exactly the same thing

They're not exactly the same thing. A GIN Index might contain a B-Tree ("balanced tree", not to be confused with "binary tree") but how the index is structured and searched is completely different.

A standard B-Tree index maps field values to record IDs, but a GIN index maps *terms* (bits of field values like words or ngrams) to a *list* of record IDs. This structure known as an "Inverted index" hence the name "Generic INverted index".

For example, a B-Tree index performs this mapping (to keep it simple, I won't bother to try to structure this like an actual B-Tree. The tuples that are indexed are the important part):

```
Hello World => 1
Hello Bertrand => 2
```

And a GIN index performs this mapping:

```
Hello => 1, 2
World => 1
Betrand => 2
```

You can see how postgres generates the inverted index by printing a ``tsvector`` to the terminal, this is a mini inverted index that maps terms to their positions/weights inside the field.

> By creating a field of what Elasticsearch calls edge Ngrams, you actually create exactly what’s inside an index, except that it’s not sorted since it’s not an index, so it will be way slower to search for your term than if you searched directly in the original database columns, which is of course way slower than searching in a unified index.

I think you're misunderstanding how GIN indices are searched.

A GIN index is searched with the following steps:

1) Break the query up into terms
2) Find each term in the index
3) Combine the posting lists for all the terms you found (using the specified boolean operator)
4) Fetch the records that are in the final posting list

Ngrams optimise steps 2 and 3 because instead of having to find all the terms that begin with "a" and combine all their posting lists on the fly (which could be extremely inefficient, as there may be thousands of terms that match that) we instead pregenerate a posting list for each prefix. The way we get postgres to generate these prefix posting lists is by indexing the edgengrams.

This is actually the recommended approach in Elasticsearch, which is a database that sits on top of an inverted index: https://www.elastic.co/guide/en/elasticsearch/guide/current/_ngrams_for_partial_matching.html#_ngrams_for_partial_matching

> Just creating a column compiling all partial_matched SearchFields without this tokenization but with a GIN index would be way faster for roughly the same table size.

Nope, that would be way slower and the index would be much smaller.

> But again, this is far from ideal compared to what we aim to achieve: having the same search functionalities with both Elasticsearch & PostgreSQL. If we want to be able to search for a single column or method, with or without partial match, the current PostgreSQL system I made is not enough.

What I just proposed is exactly how this is done in Elasticsearch.

**Comment by BertrandBordage** [comment]
No, the GIN index does not break down data into terms. It’s `to_tsvector` or `plainto_tsvector` that do this.

Anyway, instead of discussing about who’s right and who’s wrong, I made [this Jupyter Notebook](https://gist.github.com/BertrandBordage/4aee04f6931e4fa8c2c351eb7eeb8e31) creating both approach: what I understand from yours, and mine.

It generates a database of absurd random English phrases. This same dataset is put in a table, and an autocomplete search is done using our approaches. As you can see, the results fetched are exactly the same. But the amount of disk is totally different while the time taken is roughly the same. Increase the size of each row from the current few words to a few hundreds of words, and you’ll see the database size grow exponentially with your approach.

But again, maybe I misinterpreted your approach, so please correct me on what I wrote in this tiny benchmark :)

Note: I changed the minimum Ngram size to 2 instead of 3 to match PostgreSQL’s behaviour that allows prefix matching starting 2 characters.

**Comment by kaedroho** [comment]
Interesting! That benchmark does look correct and if I checked it correctly, it does prove me incorrect :)

There must be something clever that Postgres is doing to get the prefix match to actually run faster than the precomputed edgengrams. It would be interesting to find out what that is and whether it also applies to Elasticsearch (it won't be the first time I've come across a completely incorrect reccomendation in their docs).

Thanks for looking into this! I'm happy to go with your approach based on the fact that it's more efficient on disk space.

**Comment by kaedroho** [comment]
Actually, one thing that might be interesting to try in the benchmark would be to add an extra term that begins with "engl" to each row. This would show us how postgres performs when combining large postings lists (which it will have to do if used in the real world). Maybe also try searching for "en" as "engl" only prefix-matches 12 terms in the dictionary which might explain why the merging on the fly is so fast.

**Comment by BertrandBordage** [comment]
I made [a new revision of the benchmark](https://gist.github.com/BertrandBordage/4aee04f6931e4fa8c2c351eb7eeb8e31), with way more data, as well as doing multiple searches like you asked. I used the word “therefore” because the prefixes “th”, “the”… match much more results than with “English”.

As you can see, with an index the results are roughly the same in terms of searching speed.
In this run your approach is 14% faster for searching, but the database size is 3.5× bigger and data takes 2.8× more time to insert.

Something I didn’t mention earlier about your approach too: your tokenisation does not take search configuration rules into account. This can make search behave weirdly, in lots of _scenarii_. For example, if your configuration contains a thesaurus that replaces “TV set” by “television”, you will have indexed “tv”, “se” & “set”, which will not be replaced by “television” as it doesn’t match the thesaurus key. If a user searches for “telev”, he will not find the result, while he would have using prefix matching.
I think there are also cases involving non-whitespace separators, stemming and synonyms.

**Comment by thibaudcolas** [comment]
I’m not exactly sure of the history here but I believe autocomplete support across DB search shipped either before, or as part of the Database search backend added in Wagtail 2.15? Either way this is definitely supported now. Closing. I’m sure there are more improvements needed, they probably warrant their own issue.
